### ISSUE TITLE
Post Processing Failed/UnicodeDecodeError

### ISSUE THREAD

**Issue by vinzz73** [sabnzbd/sabnzbd#570]
I get the following error on certain spots:

:ERROR::[postproc:523] Post Processing Failed for ...
UnicodeDecodeError: 'utf8' codec can't decode byte 0xfc in position 10: invalid start byte

Sabnzbd version:
git status
On branch develop
Your branch is up-to-date with 'origin/develop'.

**Comment by shypike** [comment]
Not enough info, I'm afraid.
Operating system?
Example NZB?

**Comment by vinzz73** [comment]
Debian Wheezy
PHP 5.6.19
NZB example, see file..

[Miscellaneous Magazines 11.zip](https://github.com/sabnzbd/sabnzbd/files/255016/Miscellaneous.Magazines.11.zip)

**Comment by shypike** [comment]
Please check this issue: https://github.com/sabnzbd/sabnzbd/issues/543

**Comment by sanderjo** [comment]
As of SABnzbd 1.0.2, there is a special log line for the issue Shypike is reffering too:

```
$ cat ~/.sabnzbd/logs/sabnzbd.log | grep -i encoding

2016-05-07 13:17:01,709::INFO::[sabnzbdplus:1262] Preferred encoding = UTF-8
2016-05-08 12:10:51,045::INFO::[sabnzbdplus:1262] Preferred encoding = UTF-8
2016-05-08 12:27:11,153::INFO::[sabnzbdplus:1262] Preferred encoding = UTF-8
```

So ... what does your sabnzbd.log say about "Preferred encoding"? On Linux it should be something with "UTF-8", @vinzz73 ?

**Comment by vinzz73** [comment]
Strange:

INFO::[SABnzbd:1345] Preferred encoding = UTF-8
locale:
LANG=en_US.UTF-8

Also the test download nzb with german and korean langauge in #543 I can download with no issues..

I don't get then..

**Comment by sanderjo** [comment]
> INFO::[SABnzbd:1345] Preferred encoding = UTF-8
> Also the test nzb with german and korean langauge in #543 I can download with no issues..

Wow ... interesting. 

One difference I can see is that 'your' post contains a ZIP-file. Shall I create a test-post with a ZIP-file in it containing the German and Korean files to verify that hypothesis?

**Comment by vinzz73** [comment]
I zipped that file myself only for Git to be able to upload it here.

**Comment by sanderjo** [comment]
No, inside the rar set is yet another zip file (size 670 MB):

`675582211 May  6 12:21 bbbb653566346445445454555554.zip`

That was the zip I was referring to.

**Comment by vinzz73** [comment]
I see, shall we test a new nzb with zip included then?

**Comment by sanderjo** [comment]
Post with special character filenames inside a zip:
https://www.binsearch.info/index.php?q=UOIUOIUOSUOSSSS

**Comment by vinzz73** [comment]
Yep also succesfully downloaded:

Name
UOIUOIUOSUOSSSS
Completed
10/05/2016 19:23
Status
Completed
Size
3 MB

**Comment by sanderjo** [comment]
Two more posts (that will appear in 15 minutes or so):

Russian in a zip: https://nzbindex.com/search/?q=russian+UOIUKSKISKSISKIS

Russian text in filename (not a zip): https://nzbindex.com/search/?q=Russian+OSIOISOIOSIOSIOS

**Comment by vinzz73** [comment]
Both succesfully downloaded

**Comment by sanderjo** [comment]
Strange that the Russian-in-a-zip does not appear on NZBindex. Here's it on Binsearch:

https://www.binsearch.info/index.php?q=russian+UOIUKSKISKSISKIS

**Comment by vinzz73** [comment]
Thanks, and also succesful

**Comment by sanderjo** [comment]
Well, then I'm out of ideas, let alone a solution. :-(

**Comment by thezoggy** [comment]
@sanderjo https://nzbindex.com/search/?q=&age=&max=25&g[]=16&minage=&sort=agedesc&minsize=&maxsize=&dq=&poster=gekkehenk%40gmail.com&nfo=&hidespam=0&more=1

**Comment by vinzz73** [comment]
I installed a new Debian vm and cloned sabnzbd from scratch. Same error. I don't get it.

**Comment by sanderjo** [comment]
> I installed a new Debian vm and cloned sabnzbd from scratch. Same error. I don't get it.

A VM? Are the two SAB Folder Temp (Incomplete) and Complete within the VM, or outside it?

**Comment by vinzz73** [comment]
Everything is stored on the same VM. Different partition though.

**Comment by sanderjo** [comment]
> Everything is stored on the same VM. Different partition though.

Is there enough space on the first partition (I mean: the partition where Debian itself is installed) to have SAB's Temp aka Incomplete **and** Complete there so? So in your case 2 * 750 = 1500 MB?
If so, could you try that?

Furthermore: what types are the filesystems on the different partitions? Here's how to check:
```
 $ mount | grep -e "^/"
/dev/sda1 on / type ext4 (rw,errors=remount-ro)
/dev/sda6 on /home type ext4 (rw)
```

**Comment by vinzz73** [comment]
`/dev/mapper/storage-storage01 on /storage type ext4 (rw,relatime,data=ordered)`

It is is an LVM volume. The test VM was also installed using LVM out of habit. Could LVM be the problem? Then again, Synology also uses LVM I believe.

`-rw-r--r-- 1 root root 0 May 12 10:58 /storage/files/incoming/Chinese-中國是非常大的-text.txt
-rw-r--r-- 1 root root 0 May 12 10:58 /storage/files/<snip>/Chinese-中國是非常大的-text.txt
`

**Comment by sanderjo** [comment]
I can reproduce it after creating a Debian 8.4 VM with LVM. That's a first step, isn't it?

BTW: the result is that the download is there, but still in a _UNPACK folder, right?

```
sander@debian:~/Downloads/complete$ ls -al _UNPACK_Miscellaneous\ Magazines\ 11/ | cut -c-60
total 714388
drwxr-xr-x 2 sander sander      4096 May 13 18:01 .
drwxr-xr-x 3 sander sander      4096 May 13 18:01 ..
-rw-r--r-- 1 sander sander  68940273 May  5 10:14 CamperTrai
-rw-r--r-- 1 sander sander  18782910 May  1 10:32 CarTravel 
-rw-r--r-- 1 sander sander  75821208 May  5 08:43 conde-nast
-rw-r--r-- 1 sander sander  50867557 May  3 14:06 Disc0v3rIn
-rw-r--r-- 1 sander sander  46097332 May  4 10:38 Discover_B
-rw-r--r-- 1 sander sander  15141424 Apr 30 17:37 Fodors_New
-rw-r--r-- 1 sander sander  16252405 May  6 08:04 Grand Tour
-rw-r--r-- 1 sander sander  11980449 May  6 07:39 Lonely Pla
-rw-r--r-- 1 sander sander  41002591 May  5 12:37 LonelyPlan
-rw-r--r-- 1 sander sander  21590490 Jan  1 21:48 Michael M
-rw-r--r-- 1 sander sander 106658018 Apr 30 09:11 Outdoor - 
-rw-r--r-- 1 sander sander 103963470 May  5 16:59 Practical_
-rw-r--r-- 1 sander sander  19031873 May  4 09:30 Sehnsucht 
-rw-r--r-- 1 sander sander  48486849 May  3 14:18 Travel_Ire
-rw-r--r-- 1 sander sander  42119137 Apr 30 11:57 VW Camper 
-rw-r--r-- 1 sander sander  44756119 Apr 30 11:59 VW CAMPER 
```


```

2016-05-13 18:01:44,937::INFO::[newsunpack:243] Unzip finished on /home/sander/Downloads/incomplete/Miscellaneous Magazines 11
2016-05-13 18:01:44,939::ERROR::[postproc:519] Post Processing Failed for Miscellaneous Magazines 11 ()
2016-05-13 18:01:44,940::INFO::[postproc:521] Traceback: 
Traceback (most recent call last):
  File "/usr/share/sabnzbdplus/sabnzbd/postproc.py", line 355, in process_job
    unpack_error, newfiles = unpack_magic(nzo, short_path(workdir), short_complete, flag_delete, one_folder, (), (), (), (), ())
  File "/usr/share/sabnzbdplus/sabnzbd/newsunpack.py", line 270, in unpack_magic
    xjoinables, xzips, xrars, xsevens, xts, depth)
  File "/usr/share/sabnzbdplus/sabnzbd/newsunpack.py", line 270, in unpack_magic
    xjoinables, xzips, xrars, xsevens, xts, depth)
  File "/usr/share/sabnzbdplus/sabnzbd/newsunpack.py", line 208, in unpack_magic
    xjoinables, xzips, xrars, xsevens, xts = build_filelists(workdir, workdir_complete)
  File "/usr/share/sabnzbdplus/sabnzbd/newsunpack.py", line 1616, in build_filelists
    for root, dirs, files in os.walk(workdir_complete):
  File "/usr/lib/python2.7/os.py", line 286, in walk
    if isdir(join(top, name)):
  File "/usr/lib/python2.7/posixpath.py", line 80, in join
    path += '/' + b
UnicodeDecodeError: 'utf8' codec can't decode byte 0xfc in position 10: invalid start byte
2016-05-13 18:01:45,252::INFO::[postproc:551] Cleaning up Miscellaneous Magazines 11 (keep_basic=True)
```


```
2016-05-13 18:01:29,982::INFO::[newsunpack:233] Unrar finished on /home/sander/Downloads/incomplete/Miscellaneous Magazines 11
2016-05-13 18:01:29,983::DEBUG::[newsunpack:1641] build_filelists(): joinables: []
2016-05-13 18:01:29,983::DEBUG::[newsunpack:1642] build_filelists(): zips: [u'/home/sander/Downloads/complete/_UNPACK_Miscellaneous Magazines 11/bbbb653566346445445454555554.zip']
2016-05-13 18:01:29,984::DEBUG::[newsunpack:1643] build_filelists(): rars: []
2016-05-13 18:01:29,984::DEBUG::[newsunpack:1644] build_filelists(): 7zips: []
2016-05-13 18:01:29,985::DEBUG::[newsunpack:1645] build_filelists(): ts: []
2016-05-13 18:01:29,986::INFO::[newsunpack:240] Unzip starting on /home/sander/Downloads/incomplete/Miscellaneous Magazines 11
2016-05-13 18:01:29,986::INFO::[newsunpack:787] Starting extract on zipfile: /home/sander/Downloads/complete/_UNPACK_Miscellaneous Magazines 11/bbbb653566346445445454555554.zip 
2016-05-13 18:01:44,772::DEBUG::[newsunpack:849] unzip output: 
2016-05-13 18:01:44,774::INFO::[newsunpack:808] Deleting /home/sander/Downloads/complete/_UNPACK_Miscellaneous Magazines 11/bbbb653566346445445454555554.zip
2016-05-13 18:01:44,937::INFO::[newsunpack:243] Unzip finished on /home/sander/Downloads/incomplete/Miscellaneous Magazines 11
2016-05-13 18:01:44,939::ERROR::[postproc:519] Post Processing Failed for Miscellaneous Magazines 11 ()
2016-05-13 18:01:44,940::INFO::[postproc:521] Traceback: 
Traceback (most recent call last):
  File "/usr/share/sabnzbdplus/sabnzbd/postproc.py", line 355, in process_job
    unpack_error, newfiles = unpack_magic(nzo, short_path(workdir), short_complete, flag_delete, one_folder, (), (), (), (), ())
  File "/usr/share/sabnzbdplus/sabnzbd/newsunpack.py", line 270, in unpack_magic
    xjoinables, xzips, xrars, xsevens, xts, depth)
  File "/usr/share/sabnzbdplus/sabnzbd/newsunpack.py", line 270, in unpack_magic
    xjoinables, xzips, xrars, xsevens, xts, depth)
  File "/usr/share/sabnzbdplus/sabnzbd/newsunpack.py", line 208, in unpack_magic
    xjoinables, xzips, xrars, xsevens, xts = build_filelists(workdir, workdir_complete)
  File "/usr/share/sabnzbdplus/sabnzbd/newsunpack.py", line 1616, in build_filelists
    for root, dirs, files in os.walk(workdir_complete):
  File "/usr/lib/python2.7/os.py", line 286, in walk
    if isdir(join(top, name)):
  File "/usr/lib/python2.7/posixpath.py", line 80, in join
    path += '/' + b
UnicodeDecodeError: 'utf8' codec can't decode byte 0xfc in position 10: invalid start byte
2016-05-13 18:01:45,252::INFO::[postproc:551] Cleaning up Miscellaneous Magazines 11 (keep_basic=True)
2016-05-13 18:01:45,253::INFO::[__init__:932] /home/sander/Downloads/incomplete/Miscellaneous Magazines 11/__ADMIN__/SABnzbd_nzf_AwWfjG removed
2016-05-13 18:01:45,253::INFO::[__init__:932] /home/sander/Downloads/incomplete/Miscellaneous Magazines 11/__ADMIN__/SABnzbd_nzf_hAOvvu removed
2016-05-13 18:01:45,253::INFO::[__init__:932] /home/sander/Downloads/incomplete/Miscellaneous Magazines 11/__ADMIN__/SABnzbd_nzf_6baLIN removed
2016-05-13 18:01:45,254::INFO::[__init__:932] /home/sander/Downloads/incomplete/Miscellaneous Magazines 11/__ADMIN__/SABnzbd_nzf_6Jf0dC removed
2016-05-13 18:01:45,254::INFO::[__init__:932] /home/sander/Downloads/incomplete/Miscellaneous Magazines 11/__ADMIN__/SABnzbd_nzf_oaTh1T removed
2016-05-13 18:01:45,254::INFO::[__init__:932] /home/sander/Downloads/incomplete/Miscellaneous Magazines 11/__ADMIN__/SABnzbd_nzf_KJra6Q removed
2016-05-13 18:01:45,255::INFO::[__init__:932] /home/sander/Downloads/incomplete/Miscellaneous Magazines 11/__ADMIN__/SABnzbd_nzf_lfer_5 removed
2016-05-13 18:01:45,256::INFO::[misc:1285] Cannot remove folder /home/sander/Downloads/incomplete/Miscellaneous Magazines 11/__ADMIN__
2016-05-13 18:01:45,257::INFO::[misc:1285] Cannot remove folder /home/sander/Downloads/incomplete/Miscellaneous Magazines 11/__ADMIN__
2016-05-13 18:01:45,257::INFO::[postproc:83] Saving postproc queue
2016-05-13 18:01:45,258::INFO::[__init__:941] Saving data for postproc2.sab in /home/sander/.sabnzbd/admin/postproc2.sab
2016-05-13 18:01:45,258::INFO::[downloader:277] Post-processing finished, resuming download
2016-05-13 18:01:46,273::INFO::[nzbqueue:250] Saving queue
2016-05-13 18:01:46,274::INFO::[__init__:941] Saving data for queue10.sab in /home/sander/.sabnzbd/admin/queue10.sab
2016-05-13 18:01:46,276::INFO::[__init__:941] Saving data for totals10.sab in /home/sander/.sabnzbd/admin/totals10.sab
2016-05-13 18:01:46,278::INFO::[__init__:941] Saving data for rss_data.sab in /home/sander/.sabnzbd/admin/rss_data.sab
2016-05-13 18:01:46,279::INFO::[__init__:941] Saving data for watched_data2.sab in /home/sander/.sabnzbd/admin/watched_data2.sab
2016-05-13 18:01:46,281::INFO::[postproc:83] Saving postproc queue
2016-05-13 18:01:46,282::INFO::[__init__:941] Saving data for postproc2.sab in /home/sander/.sabnzbd/admin/postproc2.sab
2016-05-13 18:01:46,282::INFO::[postproc:680] Queue has finished, launching: None (None)
```

**Comment by sanderjo** [comment]
Googling the error
```
  File "/usr/lib/python2.7/posixpath.py", line 80, in join
    path += '/' + b
```
gives interesting: Always with Unicode (mixing with non-Unicode filenames), and even sometimes in combination with a zip file.

TBC

**Comment by vinzz73** [comment]
Thats something indeed. Yes, the  _UNPACK folder remains. Will install of Python 3.4 solve this? I now use 2.7.

**Comment by thezoggy** [comment]
no. sab is written in python 2 syntax not 3. thus python 3.x is not supported. you could just think of it as a different language :)

**Comment by sanderjo** [comment]
Some technical notes (ignore if you want):

On Debian:

files ['..., 'Michael M\xfcller - Dresbla', ...]
file is Michael M�ller - Dresbla, which can be decoded as cp1252

On Ubuntu:

files [ ...', 'Michael M\xd0\x91ller - Dresbla', ...]
file is Michael MБller - Dresbla, which can be decoded as utf8

Note: the above files are the result of the unzip; they were posted in a ZIP-file. Maybe there is something going wrong in SAB or Debian ... maybe not forced into UTF-8 format. 

Maybe the poster has posted the ü letter in the filename as something else than UTF-8 (possibly indeed cp1252, which is used on Windows)


Another possibility ( @shypike ?): ignore the UnicodeDecodeError error as the files are just there ... ?

**Comment by sanderjo** [comment]
@vinzz73  on your Debian, can you download this NZB with SABnzbd: https://nzbindex.com/search/?q=herr+Windows+file+with+umlaut+in+a+zip , and post back the result?

**Comment by sanderjo** [comment]
After a deep dive, I found the root cause: Windows and Zip-files can encode special characters in Code Page 437 (cp437). All 8-bit. That's OK.

However, on Linux such a zip file is unzipped as Unicode, which is wrong; the Linux unzipper should look at bit 11 of "General purpose bit flag": if that bit is 0, the filenames inside are ASCII / CP437. 
In case of special characters (thus CP437, not plain ASCII), unzipping it directly as Unicode is wrong and gives the error that started this thread.

I wrote a program to analyse zip files regarding the encoding. See https://github.com/sanderjo/zipfile_encoding_info

I'm not sure SABnzbd should solve this. However,I do think SABnzbd should be aware of the situation and shoud not give a traceback.

**Comment by vinzz73** [comment]
This option maybe?

`The option -UU allows to entirely disable the recognition of UTF-8 encoded filenames.`

**Comment by sanderjo** [comment]
@vinzz73 

On https://github.com/sanderjo/zipfile_encoding_info there are a two zip files created on Windows. Can you if "unzip ... -UU ..." creates correct German names in the unpacked files?
I tried, but that did not work on Ubuntu.

**Comment by sanderjo** [comment]
This is a proof of concept how SABnzbd gets the traceback:

```
#!/usr/bin/python
# -*- coding: utf-8 -*-

import os
import sys
import zipfile

basedir = u"blabla"
inputfile = sys.argv[1]

zip=zipfile.ZipFile(inputfile)
files = zip.namelist()
print files

for _file in files:
	print "join is", os.path.join(basedir, _file)

```

A traceback will occur if the zip-input-file is Windows-encoded with special CP437 characters inside. Examples:

```
$ python let-it-crash.py ein-Herr-zu-hause---Windows.zip 
['Herr-M\x81ller-ist-zu-hause.txt.txt']
join is
Traceback (most recent call last):
  File "let-it-crash.py", line 16, in <module>
    print "join is", os.path.join(basedir, _file)
  File "/usr/lib/python2.7/posixpath.py", line 80, in join
    path += '/' + b
UnicodeDecodeError: 'ascii' codec can't decode byte 0x81 in position 7: ordinal not in range(128)

```
and

```
$ python let-it-crash.py "/home/sander/Downloads/complete/Miscellaneous Magazines 11/bbbb653566346445445454555554.zip"
['Fodors_New_York_City_2016.epub', 'Michael M\x81ller - Dresden.epub', ...]

file is Fodors_New_York_City_2016.epub
join is blabla/Fodors_New_York_City_2016.epub
file is Michael M�ller - Dresden.epub
join is
Traceback (most recent call last):
  File "let-it-crash.py", line 17, in <module>
    print "join is", os.path.join(basedir, _file)
  File "/usr/lib/python2.7/posixpath.py", line 80, in join
    path += '/' + b
UnicodeDecodeError: 'ascii' codec can't decode byte 0x81 in position 10: ordinal not in range(128)

```

**Comment by sanderjo** [comment]
And with a few extra lines of code, no more crashes:

```
#!/usr/bin/python
# -*- coding: utf-8 -*-

import os
import sys
import zipfile

basedir = u"blabla"
inputfile = sys.argv[1]

zip=zipfile.ZipFile(inputfile)
files = zip.namelist()
print files

for _file in files:
	print "file is", _file
	try:
		print "join is", os.path.join(basedir, _file)
	except:
		# the encoding was a problem, so let's assume it was a zip with special CP437 characters in the filename
		print "join is", os.path.join(basedir, _file.decode('cp437'))

```

```
$ python let-it-crash.py twee-files-met-duitse-tekens---Windows.zip 
['Deutsch-ist-gef\x81hlsm\x84\xe1ig-jajaja.txt', 'Herr-M\x81ller-ist-zu-hause.txt.txt']
file is Deutsch-ist-gef�hlsm��ig-jajaja.txt
join is join is blabla/Deutsch-ist-gefühlsmäßig-jajaja.txt
file is Herr-M�ller-ist-zu-hause.txt.txt
join is join is blabla/Herr-Müller-ist-zu-hause.txt.txt
```

```
$ python let-it-crash.py /home/sander/git/zipfile_encoding_info/Chinese-filename-inside-by-7zip.zip
[u'Chinese-\u6211\u559c\u6b22\u5f53\u4e00\u4e2a\u8ba1\u5212\u662f\u4e00\u8d77-plan.txt']
file is Chinese-我喜欢当一个计划是一起-plan.txt
join is blabla/Chinese-我喜欢当一个计划是一起-plan.txt
```

**Comment by shypike** [comment]
Any suggestions for a fix in SABnzbd?

**Comment by sanderjo** [comment]
Yes.

In the code above, the solution is
```
zip=zipfile.ZipFile(inputfile)
files = zip.namelist()
print files

    try:
        print "join is", os.path.join(basedir, _file)
    except:
        # the encoding was a problem, so let's assume it was a zip with special CP437 characters in the filename
        print "join is", os.path.join(basedir, _file.decode('cp437'))
```
So ... SAB should do the same. Somewhere in the SAB code (I don't know where) SAB is parsing over the files in the zip file, and doing a join(). Can that be true? If so @shypike , can you give a pointer? I can make to try a fix in SAB.

**Comment by shypike** [comment]
Actually, SABnzbd doesn't parse the filenames inside the zip. It just does a basic "unzip <zip-name>".
The crash seems to occur in newsunpack.build_filelists(). This is called after the ZIP file has been unpacked, to scan for nested archives to be unpacked.
The crash is therefore triggered by what comes back from the operating system, not from unzip.
Some parts of the Python library aren't very good at handling Unicode though.
I'll need to go over this with a debugger.

**Comment by sanderjo** [comment]
Maybe just ignore the problem with a try/except/pass so that SAB can go on?

Reason: it's a very niche problem. It occurs only if the poster posted a (zipped?) post with a non-US-ASCII-character (for example: ü or é, so the European special characters ) AND it was not encoded with Unicode/UTF-8 (but with cp1252 or so)

This is the problem: first command is OK (just encoding in cp1252), second command goes wrong (trying to concat a cp1252-encoded string with a unicode-encoded string):

```
>>> print u"müller".encode('cp1252')
m�ller

>>> print u"müller".encode('cp1252') + u"blabla"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'ascii' codec can't decode byte 0xfc in position 1: ordinal not in range(128)
```

That can't happen with Asian characters, as they don't occur in cp1252

**Comment by shypike** [comment]
If I just knew where.

**Comment by sanderjo** [comment]
It's this code, isn't it:
```
    if workdir_complete:
        for root, dirs, files in os.walk(workdir_complete):
            for _file in files:
                if '.AppleDouble' not in root and '.DS_Store' not in root:
                    filelist.append(os.path.join(root, _file))
```
What would happen if we put a try/except around, and thus the result would be that ```filelist``` is not or only partially filled?

**Comment by shypike** [comment]
OK, makes sense.
Problematic names will be skipped. Which isn't a problem as long as a ZIP or RAR file doesn't have the problematic name. In the latter case, they will not be unpacked.

**Comment by shypike** [comment]
OK, this should be the fix: abe35f91004012080c1ec66649be75a9eabf6f09

**Comment by sanderjo** [comment]
I think you mean https://github.com/sabnzbd/sabnzbd/commit/6108f9c703462fb3a1481a44ecfde3be6336ad4a

**Comment by shypike** [comment]
Yeah!

**Comment by shypike** [comment]
@sanderjo @vinzz73 Close this one. Should be solved in 1.1.0RC4.
